### ISSUE TITLE
Decoding issue although TTS training apparently successful

### ISSUE THREAD

**Issue by siiyouaguein** [espnet/espnet#3451]
Greetings,

I am working on a custom recipe to train a TTS model using the english Common Voice dataset in ESPnet 2. This recipe was based on data preprocessing from the Common Voice ASR recipe and on the training recipe from LibriTTS. I have added a custom silence trimming process in order to deal with silences at the beginning and the end of audios, and selected only data from the _validated_ set. The text is phonemized using g2p_en, and I have not touched the phonemes. 

As I am relatively new to TTS, I have had many bugs to deal with, and I would like to thank you for the different tutorials and answers to issues available.  
But I am currently facing an issue I believe is linked somehow to the decoder. Indeed, after training my model with Tacotron2 for about 80k iterations, I get acceptable attention graphs and stabilizing losses:
![attn_loss](https://user-images.githubusercontent.com/73817271/129400808-308e4601-d2d9-40df-9b8a-68236a617510.png)
![loss](https://user-images.githubusercontent.com/73817271/129400823-b2af43a6-5b9f-41b0-94ad-2057e80796b6.png)
![tts dec att 164ep](https://user-images.githubusercontent.com/73817271/129400835-67ea606c-303a-47c9-9cbb-aafbbec6253c.png)
![tts dec att 164ep](https://user-images.githubusercontent.com/73817271/129400845-647997c4-8f90-4547-ae32-e27cd5229800.png)

FYI, the splittings in the middle appeared a few epochs before this curve. The previous curves do not have these flaws.

But as I proceed to stage 7, I get completely strange noise. I searched for many possible reasons for this, but I believe there is an issue with decoding, as I get these attention weights:
![0027fcb0b1be4cd41e38167393b6dff0ba22b28b4bb450aa790774213ef332c59e431d93c193c8a7853ccae394fea55a365840f2ba950d0ef15101d71674cd1f-common_voice_en_20911758](https://user-images.githubusercontent.com/73817271/129401342-f9e4533f-e44d-4de4-aed3-f75e4f5e8f02.png)
[decoding_output_example.zip](https://github.com/espnet/espnet/files/6984407/decoding_output_example.zip)

Do you have any idea as to why this happens? 
I add my config and the decoder YAML file below. Thank you in advance, I previously have tried to solve my problems as independently as possible but this particular problem troubles me.

I use Linux OS on a Docker container, with a GTX 1080 GPU.
[configs.zip](https://github.com/espnet/espnet/files/6984434/configs.zip)

**Comment by kan-bayashi** [comment]
At first could you try `use_att_constraint=false` in decode.yaml?
This may give us more hint.

**Comment by siiyouaguein** [comment]
Thanks for the reply.
After changing att_constraint, the results were the same. 
I notcied that for example, the durations file show a pattern of stopping at around a third of the alignment job:
[durations.txt](https://github.com/espnet/espnet/files/6986087/durations.txt)
which is consistent with the attentions graphs I get. Maybe for some reason the process stops, or it does not manage to read the stop tokens correctly during decoding ?

**Comment by kan-bayashi** [comment]
Hmm. I noticed that attention loss is high (I checked my case, which is around 5e-5).
Now I'm wondering the training is totally failed but the first attention plot seems fine, which is strange if the training is failed.
How did you get the attention plot of the first figure? (from `exp/train*/att_ws`)?
And could you show me one of the training data (audio waveform and text)?

**Comment by siiyouaguein** [comment]
The first two figures, where the attention is globally diagonal, come from the training phase, from att_ws.
I selected a random audio file from validated.tsv:

All the species bear yellow flowers similar to small sunflowers.
[common_voice_en_20990020.zip](https://github.com/espnet/espnet/files/6986414/common_voice_en_20990020.zip)
![image](https://user-images.githubusercontent.com/73817271/129445569-449e875c-195d-49ab-aad8-8aa329212751.png)
![image](https://user-images.githubusercontent.com/73817271/129445577-0890e310-f4aa-477f-a64b-0cf657fcd3de.png)


I know that some of the audios in this dataset is noisy, or not loud enough, but the purpose of my experiment was to determine if the data size could compensate these imperfectinos during training, which is why I would like to determine if the problem truly comes from the decoder or from the data (possibility that some corrupted audio files still present in vatidated.tsv)

**Comment by sw005320** [comment]
I like your motivation.
Actually, we are working on similar projects (use ASR data for TTS), and we may collaborate.
In addition to the current efforts, we may use
- speech enhancement (to denoise speech or compare the original and enhanced speech and prune the utterances that are very different)
- CTC scoring (to detect the label mistake)

**Comment by kan-bayashi** [comment]
Thank you for sharing your motivation.
In the case of AR model, trimming the silence at the beginning and the end of audio is very important even if the audio is clean.
You can try simple trimming script like this:
https://github.com/espnet/espnet/blob/4406f25ebf507daf33f68787ff0e3699a0937913/egs2/kss/tts1/local/data.sh#L72-L83

**Comment by siiyouaguein** [comment]
Thank you for the code snippet.
I saw that you adapted the trim_silence code for other recipes in a recent update. As I had not managed to convert the existing code before ( ike the one from the LibriTTS recipe), I used a method based on ffmpeg silenceremove option, and then tuned the hyperparameters based on the results; the problem with this technique is that snipping silences for files too light caused corruption of data; similarly, choosing a threshold too high caused deleting a few of the words, which is why in some audios like the one I sent you there are still a few silences (actually not trimmed because the sound of the microphone being turned on and off stops the process).

I will try again using the new silence trimming you offer, and will keep you updated.

@sw005320 , my second idea was trying with GigaSpeech, but for another project I was already trying to use Common Voice due to multiple languages. But your ideas are very interesting indeed and may solve the few issues remaining with those "collaborative" datasets!

**Comment by kan-bayashi** [comment]
> choosing a threshold too high caused deleting a few of the words, which is why in some audios like the one I sent you there are still a few silences (actually not trimmed because the sound of the microphone being turned on and off stops the process).

Yeah, that is a common problem...
I have not compared strictly but aggressive trimming is better than weaker trimming especially for AR based attention modeling. If you investigate such kinds of effects, that is great:)

**Comment by siiyouaguein** [comment]
Greetings,
After a few issues (notably the creation of almost 400 Gb of data in stage 2 due to the conversion to .wav, which filled my drive), I tried the training, using the trim_silence.sh code you gave me. First results, with 20 epochs for now, give me more coherent decoding curves, and one can distinguish the first presence of audio in the generated wav files ( although very limited, seeing the early stage of training):
![00a46a27afb6e089076fa667263a69f97d00e0ad927f4d2cb777da19e73cd6a926c7665ab8a7152b1fab97d3bc4ff6db989ef3b59a274b5f19a76f92760601b1-common_voice_en_19617110](https://user-images.githubusercontent.com/73817271/130583728-e7e376d0-f6ef-4b1d-a9c5-4db0b1790ad0.png)
 
[20_epoch_audio.zip](https://github.com/espnet/espnet/files/7037699/20_epoch_audio.zip)

Thus it appears that a more aggressive trimming might have been the solution to the problem. I will try to keep you updated on the rest of training, although my internship is coming to an end so I might not see the full result. Thanks for your help!
